struct.lua

```diff
@@ -0,0 +1,202 @@
+--[[
+ * Copyright (c) 2015-2020 Iryont <https://github.com/iryont/lua-struct>
```

Comment:
I wholehartedly agree, there are two reasons I could not do this better, even if I wanted to.

1) Using submodule, forking the repository in official kaitai-io organization, setting up the dependencies or dependabot, whatever, is not something I as an external contributor can do

2) Introducing dependency management or package management requirement to the runtime is again something, I should not do, because it fundamentaly changes the way both kaitai-io and any external integration/implementation works with the runtime, and that is something only maintainers should do, because that is strategic/architectural decision.

Let's not forget the whole reason to this PR, the reason is to provide as seamless transition to compatibility with versions 5.1+ and LuaJIT as possible. Doing the things as proposed here in your comment, would break a lot of things and trigger a lot of backlash for the maintainers.
Which is why I introduced these ideas in separate tickets, that expected some future release to be "breaking changes", but not with this PR, see issues #6 and #11 for reference

So to conclude, if you want to do the breaking changes right now, either give me the permission to do so (i'm not gonna propose another thing, just to wait 2 months for your review), or at least put up the maintainer decision on what should be and should not be done (ie. give resolution to proposals and most importantly to questions in #6 and https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#issuecomment-673364448 ), so someone else can take care of it, but I'm not going to do that as external contributor, if I cannot be sure, if maintainer will approve of me doing so.

Reply:
> Please don't add dependencies by copy & pasting their source code

While I agree @generalmimon I'd say that vendoring sources is very common practice. In this case, a submodule is best IMO, but @smarek is right - you'd need to set this up as a member yourself.

Reply:
> Using submodule, forking the repository in official kaitai-io organization, (...), is not something I (...) can do

> but @smarek is right - you'd need to set this up as a member yourself.

The ownership is actually the smallest problem, every repo can be easily transferred to another GitHub user or organization. But sure, if you want me to set up the forks on the kaitai-io org straight away, here they are:

 - https://github.com/kaitai-io/lua-struct
 - https://github.com/kaitai-io/nmap

Reply:
I probably also have to create the target branch for custom changes (of course, we will **not** be polluting `master` because this branch should reflect the upstream `master` - you know the drill) needed for our use in Lua runtime. 

Is the `ks-runtime` name OK or can you come with a better name, @smarek?

Reply:
Name is ok, but also the thing that needs least of our attention right now, HOWEVER it's very nice you did those forks, I'M SO HAPPY YOU DID, i'm gonna waste few more days on changing everything in my proposal, to your newly uncovered expectations, just so you can change your mind again in a month or two. /s

Using submodule means, we need to change the deployment model, which calls directly for #6 to be addressed, or better dependency-management system to be chosen and implemented.
Also users of this runtime should switch in near-future from simple cloning the runtime and adding it to LUA_PATH, to using compiled tagged releases from github (now non-existent), because imho you don't want to impose the submodules and packaging/renaming mess onto users. It's certainly good and nice way of splitting up the development process, but it's very different from what this project is currently, and how it is currently used irl.

however @generalmimon you missed my point, which is, I'm not maintainer and I cannot do the decisions about these kind of things. I do not want the lua runtime repo transferred, I want someone who will take care and active development in it. And I'm probably not the right person (i'm least skilled Lua developer among contributors here), and @generalmimon is probably not the right person, because it takes him 2 months to address PR/patch-set, which is ready to merge, just to say that the requirements and the test-cases changed in the meantime, and the contribution is now not good enough (see https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#discussion_r497837043 )

So, as I explained here https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#discussion_r497834131 I don't want you @generalmimon to put up some forks and branches for us to contribute to. I want to see you leading the project and integrating the changes with your expectations and responsibilities, or I want you to appoint maintainer, who will do this, if you cannot/don't want to. I need you to set up the source/ internal-external api/interface architecture, development/deployment/testing model and give meaningful responses so the contributors.

>1. Using submodule, forking the repository in official kaitai-io organization, setting up the dependencies or dependabot, whatever, is not something I as an external contributor can do

**means, i cannot decide this is the way it should be organized, it can be done by maintainer, and the first step, re-organizing the project and updating all relevant parts of the ecosystem, has to be done by maintainer, not by external contributor**

----

I will not contribute by setting up LuaRocks, if I'm not sure maintainer wants it, and if there is missing decision about luarocks.org publishing account ownership (see #6)
I will not ex-post re-write my entire contribution multiple times, and re-test it, just because the expectations were not clearly defined before, or maintainer just figured it could be done (see https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#discussion_r495541909 )
I will not fix up the mess around handling large numbers, unless I get maintainer decision, how it should be done (see https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#issuecomment-673364448 )
I will not replace pure-lua with compiled extensions, just because it would be less work to maintain it (see https://github.com/kaitai-io/kaitai_struct_lua_runtime/pull/5#discussion_r497821291 )

And I will certainly not put more effort to participating in this project, if all I get is long silence followed by nagging and nitpicking, if I wanted it like that, I would've hooked up with my ex-girlfriend.

Reply:
> to your newly uncovered expectations, just so you can change your mind again in a month or two. /s

There is absolutely no need for the snark. Let's keep things civil, please.

>  because it takes him 2 months to address PR/patch-set

This is OSS, not a full time job. We do this on our own time, for free, out of passion. Sometimes, we're unable to address things in a timely manner (I don't mean to speak for @generalmimon but I assume the situation is the same). Sometimes it's less friction to simply not respond to an issue than to only be able to comment but take no action due to time constraints. It's not malice nor negligence, so please don't assume as such. 

> I do not want the lua runtime repo transferred, I want someone who will take care and active development in it.

Then why not use the submodule directly from the source maintainer? There's no requirement for dependabot; only if it's necessary to update or someone requests it. Submodule commits are pinned anyway, so the code won't break without manual intervention.

> just to say that the requirements and the test-cases changed in the meantime

I'm pretty sure a number of people, not just @generalmimon, have raised some valid concerns about the approach. The work is appreciated, but if this is such a time-sensitive matter for you then the usual route is to fork and maintain on your own until the upstream maintainers are ready to make changes.

This is pretty common in OSS. We're all human, and sometimes we need time think about things too. You can't expect everyone to be on your schedule.

> I want to see you leading the project and integrating the changes with your expectations and responsibilities, or I want you to appoint maintainer, who will do this, if you cannot/don't want to. I need you to set up the source/ internal-external api/interface architecture, development/deployment/testing model and give meaningful responses so the contributors.

This is all incredibly unreasonable. You're asking for vague, unactionable work items that have very little relevance to _this_ PR. You're making harsh demands and asserting yourself as "correct" when in reality this just isn't how open source works.

So far @generalmimon has given _very meaningful responses_ to all of our concerns. Perhaps they're not the answers _you_ were hoping for, but they were certainly meaningful and were made with the best intentions for the project.

> And I will certainly not put more effort to participating in this project, if all I get is long silence followed by nagging and nitpicking, if I wanted it like that, I would've moved back with my ex-girlfriend.

That's fine, your attitude isn't really conducive to the conversation anyway. I'm sure someone else would be happy to contribute their time and energy with a much happier outlook 👍

Reply:
>There is absolutely no need for the snark. Let's keep things civil, please.

@Qix- Yeah, sorry about the tone, i'm salty, and I really wanted to express, that I feel the way @generalmimon sometimes treats the contributions is hostile and unappreciative, previous clashes were here https://github.com/kaitai-io/kaitai_struct_compiler/pull/206 and here https://github.com/kaitai-io/ci_targets/pull/7
See the discussion. Sometimes the points raised are valid, but [sometimes](https://github.com/kaitai-io/ci_targets/pull/7#discussion_r472473080) it's just nasty and completely futile effort to show ones power over petty details

>This is OSS, not a full time job. We do this on our own time, for free, out of passion. Sometimes, we're unable to address things in a timely manner (I don't mean to speak for @generalmimon but I assume the situation is the same). Sometimes it's less friction to simply not respond to an issue than to only be able to comment but take no action due to time constraints. It's not malice nor negligence, so please don't assume as such.

Since I've spent several days, as you correctly of my free time, making this contribution, where discussion was lively and constructive (I've discussed the changes over on Lobby and in several places here), i'm really disappointed that fulfilling the requirements I've obtained from kaitai team, seems to vanish in the light of all the new requirements i see today.

I manage and maintain several opensource projects, and I can understand there is little time to invest sometimes, but if maintainer can put out, just for me, 4-page lecture about how some parts of the contribution could be done better, and suddenly wants me to re-work it completely, it's insulting to me. As a maintainer, i'd rather merge the PR into development branch, do some changes (which takes roughly same time as writing the mentioned essay) and put up a todo-list of what must be solved before merging into master / publishing new version/tag.

>Then why not use the submodule directly from the source maintainer? There's no requirement for dependabot; only if it's necessary to update or someone requests it. Submodule commits are pinned anyway, so the code won't break without manual intervention.

I agree with using the submodules, but it's change breaking the current deployment model, and there are no tagged releases, compatibility notes, usage documentation, nothing, so I tried to keep this PR as simple as possible, and outline the breaking changes for the future contributors.
However it's breaking change, and as such, I'd rather have direct permission from maintainer to do it, or have the maintainer set-up the process. That is something RFC discussion could touch, but there was none, and I'm certain, now it's too late to expect me to change architecture of my contribution so drastically

I updated all necessary parts of kaitai projects, and raised the Lua runtime from 75 to 95% of KST adoption (just roughly guessing the numbers, since I don't really remember), updated tests and did everything I could, to prepare as smooth transition with the PR for existing applications/integrations, as possible.

> I'm pretty sure a number of people, not just @generalmimon, have raised some valid concerns about the approach. The work is appreciated, but if this is such a time-sensitive matter for you then the usual route is to fork and maintain on your own until the upstream maintainers are ready to make changes.
>This is pretty common in OSS. We're all human, and sometimes we need time think about things too. You can't expect everyone to be on your schedule.

Not really time-sensitive, i'd have done exactly what you describe in that case, but I can see the community is active, and I suspect the Lua is not very high-profile target for the team, so if I as a maintainer would receive contribution in marginal part of my project, i'd be quick to get it tested and merged, so the things can move on, and I motivate contributors, to actually try to contribute next time. And what I'm seeing here is the opposite.
I just cannot wrap my head around maintainer having enough time to write essays, and not enough time to test and merge contributions, that might actually push the project forwards.

>This is all incredibly unreasonable, pushy, and frankly unbelievably rude. You're asking for vague, unactionable work items that have very little relevance to this PR. You're making harsh demands and asserting yourself as "correct" when in reality this just isn't how open source works.
> So far @generalmimon has given very meaningful responses to all of our concerns. Perhaps they're not the answers you were hoping for, but they were certainly meaningful and were made with the best intentions for the project.

I'd say reasonable, because if I see the project / community / contributors crippled by maintainer in-activity, irresolution or simple neglect, i'm disgusted and makes me regret the time invested into the community. Rude? Yes, that was the point.
What I'm asking for is for maintainer to answer questions, that are crucial to contribution process, and if maintainer ignores the questions, my ability and will to contribute goes to near zero, and if that happens only after the contribution was prepared, several times revorked, discussed with the community and tested, i feel attacked and disrespected.

I've done what I'd expected from maintainer, I've outlined what I think should be done to the project, so it has at least some release-development cycle and identified what could be done better, refactored and solved in future releases, that will have to have changes, breaking compatibility, see #6 #7 #8 #9 #10 #11 , and yet we now have discussion about me having to do all the breaking changes, before the #5 can be merged? I can assure you, I thought it was just a misunderstanding, but further into the discussion, i pick more of a hostile vibes.

> That's fine, your attitude isn't really conducive to the conversation anyway. I'm sure someone else would be happy to contribute their time and energy with a much happier outlook +1

I'm really trying to be constructive, that's why I did all the work on getting the Lua compatibility to community expectations, if you like being picked on, enjoy it, but I won't appreciate it.

Reply:
This discussion I consider solved, i'm really against doing breaking changes right now, and those should be done in future by addressing #6 and #11 together with using the submodules, as proposed by @generalmimon 
This patch is about adding the compatibility with Lua 5.1+ as smoothly as possible, and that is doable only by using the vendoring sources now